AC_Avoid: active proximity avoidance

IamPete1 opened the pull request:
This adds active proximity avoidance so we can 'run away' from obstacles. 

I spent quite a long time messing with various methods to do this directly in the position controller, but I ended up back using velocity. If over the proximity margin Avoid requests the velocity that would result in stopping at the margin, the idea being that this should be quite 'soft'. However in AirSim it can be quite aggressive, I think this is mostly due to AirSim.

This also removers the pilot commanded acceleration while avoiding. Previously if you were hard against a limit with full stick when you released the stick you would spring off the limit due to acceleration feed forward. Now nothing happens, I think this is more of a pilot feel improvement than anything else. I did also play with requesting acceleration without much luck.

This has only been tested in AirSim so far, I am quite limited in the real life testing I can do at the moment so it would be great to get some feedback.

There are also issues due to the proximity sensors seeing the ground when low, and when stopping and coming back level the distance changes, the solution is to correct the proximity sensors  for roll and pitch.

rmackay9 commented:
Great to see some effort put into this, thanks!

IamPete1 commented:
https://github.com/ArduPilot/ardupilot/issues/7706

IamPete1 commented:
@rishabsingh3003  Relatively happy with the new method for avoiding, needs some more work to test exclusion/inclusion stuff and what happens if we get stuck between a fence and a proximity or some other combination. I have also broken the switch I originally added, I think its probably still a good idea to have the ability to switch it off.

IamPete1 commented:
Some testing of this PR https://www.youtube.com/watch?v=LllCOrNLWVc&feature=youtu.be

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -68,6 +68,14 @@ const AP_Param::GroupInfo AC_Avoid::var_info[] = {
+    // @Units: m/s
+    // @Range: 0 2
+    // @User: Standard
+    AP_GROUPINFO("ACTIVE_SPD", 6, AC_Avoid, _active_speed, 0.0f),
```
I wonder if it might be better to call it BACKUP_SPD or BCKAWY_SPD.  Of course the vehicle is not always moving in reverse but I think it might be easier for users to guess what it is for.

IamPete1 replied:
This actually doesn't do anything anymore, it currently just uses the loiter speeds I guess, might be a good idea to re-instate this tho.

rmackay9 replied:
Yes, I think it would be good to set a maximum backaway speed.  I think about 50cm/s would be a good default value.  My worry is just that there is some kind of sensor error especially when the vehicle is just taking off that could lead to the vehicle flying in a random direction potentially towards the pilot.

IamPete1 replied:
done, also improved name

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -274,6 +274,22 @@ void AC_Avoid::limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel_
     }
 }
 
+void AC_Avoid::limit_accel_velocity(float kP, float accel_cmss, Vector2f &desired_vel_cms, Vector2f &desired_accel_cms, const Vector2f& limit_direction, float limit_distance_cm, float dt)
```
Let's put a comment above this implementation to clarify it's purpose.

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -737,11 +753,18 @@ void AC_Avoid::adjust_velocity_polygon(float kP, float accel_cmss, Vector2f &des
+    Vector2f safe_accel(desired_accel_cmss);
+
+    // Active avoidance distances
+    Vector2f max_limits;
```
for me "max" and "limits" are pretty similar words.  Maybe it's distance_max?  not sure..

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -799,13 +831,22 @@ void AC_Avoid::adjust_velocity_polygon(float kP, float accel_cmss, Vector2f &des
         }
     }
 
+    // Active avoidance
```
maybe we can improve this comment?  "// safe location is between min and max limits".. again I'm not sure

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -847,27 +888,34 @@ float AC_Avoid::distance_to_lean_pct(float dist_m)
 
-// returns the maximum positive and negative roll and pitch percentages (in -1 ~ +1 range) based on the proximity sensor
-void AC_Avoid::get_proximity_roll_pitch_pct(float &roll_positive, float &roll_negative, float &pitch_positive, float &pitch_negative)
+// returns the roll and pitch ratios (in -1 ~ +1 range) based on the proximity sensor
```
Are they really "ratios" though?  when I read this it makes me think it's the ratio of roll to pitch.  I think they are actually roll and pitch angles expressed as a percentage from -1 to +1.

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.h:
```diff
@@ -45,8 +45,13 @@ class AC_Avoid {
      * Note: Vector3f version is for convenience and only adjusts x and y axis
      */
-    void adjust_velocity(float kP, float accel_cmss, Vector2f &desired_vel_cms, float dt);
+    void adjust_velocity(float kP, float accel_cmss, Vector2f &desired_vel_cms, Vector2f &desired_accel_cmss, Vector2f &rel_save_location_cm, float dt);
```
If possible let's explain in the comments what "rel_save_location_cm" is.  I wonder is it "safe" instead of "save"?  sorry to be annoying but perhaps let's call it safe_location_rel_cm.  I guess that it's relative to the vehicle's current target location?  or maybe relative to it's current location..?

IamPete1 replied:
yep, one of my standard typos, changed to safe_location_rel_cm

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.h:
```diff
@@ -87,6 +92,9 @@ class AC_Avoid {
     bool limits_active() const {return (AP_HAL::millis() - _last_limit_time) < AC_AVOID_ACTIVE_LIMIT_TIMEOUT_MS;};
 
+    // Enable or disable active proximity avoidance
+    void set_active_proximity(bool state) {_active_proximity = state;};
```
I think maybe we should call the feature "backup" or "backaway" instead of "active".  It's your baby so I won't press on this too much but I think for example that the simple avoidance we do in alt-hold is very "active".  that's obviously a completely different aspect of simple avoidance but I think it could be confused with this new feature so I think a more specific name would be good.

rmackay9 commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -182,6 +182,8 @@ class RC_Channel {
         Q_ASSIST =            82, // disable, enable and force Q assist
         ZIGZAG_Auto =         83, // zigzag auto switch
         AIRMODE =             84, // enable / disable airmode for copter
+        ACTIVE_PROX_AVOID =   85, // Set active proximity based avoidance
```
I'm not sure we need another RC option just for the backaway feature.  I think if the user runs into trouble they can just disable the entire simple avoidance feature.

IamPete1 replied:
removed

rmackay9 commented:
Really great to see this and special thanks for the video.  This is probably the most important feature that we need to get added to "Simple Avoidance".

Hope it's OK, I've asked @rishabsingh3003 if he can help with testing because he's also been looking at avoidance although mostly in the path planning area.

As per usual my feedback is mostly around naming and comments.

rmackay9 commented:
Ah, can we also test what happens with Rover?

IamPete1 commented:
Some testing from @rishabsingh3003 would be great! In particular combinations of fences and proximity or proximity from two sides. Hopefully it will move to the middle of the obstacles. I suspect its also backwards for stay out zones. The moving back only works in copter loiter currently, need to add and test it for other modes too.

rishabsingh3003 commented:
Hi @IamPete1 @rmackay9 
I just learnt how to animate moving objects in Blender, and then tested this PR.
Have a look: https://youtu.be/8X799VDS8GM

It takes a little bit of time to animate since I am still new to this, but ill make more scenarios.
1. Two parallel walls coming closer together swandwiching the vehicle?
2. Fence + moving walls?
Is this what you want next? You can make a detailed list, and i'll test them and post similar video.

IamPete1 commented:
@rishabsingh3003 nice vid! Yeah one fixed and one moving wall, and moving wall vs a fence.

IamPete1 commented:
I am now happy with this a far as features go. It only works in copter loiter or guided. Should be no chance to rover, or anything else. 

I will re test and make sure the back-away speed param is working as expected. 

@rishabsingh3003 If your looking for things to do, there are the separate methods for circles in AC_Avoid. Circle fence, circle inclusion and circle exclusion, would be nice to cobine them in to a circle equivalent to adjust_velocity_polygon. Circle fence and circle in inclusion are just a copy paste of each other so that would be a nice tidy up.

Would be nice to make this work in rover also, its a not so easy as its called at the calculating speed to throttle level, so we can't just ask the nav controller to go to a different position instead. We would probably not want it in all modes either as avoidance is now for rover.

rmackay9 commented:
@rishabsingh3003, txs for the testing and video.  If possible it would be good to zoom in more on the Copter.  It's so small that we can just see it's position but we can't really see how it's responding in roll and pitch.

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -68,6 +68,14 @@ const AP_Param::GroupInfo AC_Avoid::var_info[] = {
+    // @Units: m/s
+    // @Range: 0 2
+    // @User: Standard
+    AP_GROUPINFO("BK_UP_SPD", 6, AC_Avoid, _backaway_speed, 0.5f),
```
AVOID_BACKUP_SPD is exactly 16 characters.. that might be easier to understand than BK_UP.

IamPete1 replied:
fixed

rmackay9 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -261,15 +300,20 @@ void AC_Avoid::adjust_roll_pitch(float &roll, float &pitch, float veh_angle_max)
         desired_vel_cms += limit_direction*(max_speed - speed);
+
+        // also limit acceleration in direction
+        desired_accel_cms.zero();
```
I wonder what affect this has on the vehicle in the direction parallel to the wall.  For example, if the vehicle reaches the marginat an angle.  Does it instantly lose it's forward desired acceleration?  I wonder if this makes it jerk or get stuck to the wall instead of sliding.

IamPete1 replied:
I tested trying a bunch of clever things with the acceleration and I couldn't get it to do anything better. For example we currently limit the velocity to a given acceleration value but don't bother setting the defied acceleration to that value, as far as I can tell it makes no difference if you do it properly or not. 

The only thing this was added for is to remove the 'spring back'. If your holding full stick towards a obstacle but the velo is limited to zero when you let go the copter assumes it needs to decelerate so its leans back. Now it does nothing. I think is nice for the pilot, the spring back is not intuitive.

https://youtu.be/LllCOrNLWVc?t=65

vs 

https://youtu.be/LllCOrNLWVc?t=118

rmackay9 replied:
I totally agree that the spring back is not good.  It perhaps doesn't matter but it didn't always spring back.  I'm sure that crept in when Leonard rewrote the Loiter controller so the pilot had direct control of the desired acceleration.

I'd really like to see a test of what happens when the vehicle is sliding along the wall... and the moment that it first reaches the margin.

rmackay9 replied:
I think instead of setting the desired_accel_cms to zero it needs to be limited so that it doesn't project towards the margin just like the velocity vector.

IamPete1 replied:
The spring back is the accleration vector projecting away from the limit. You can fly round in loiter just fine with it zeroed all the time.

rmackay9 replied:
It's hard to believe that simply zeroing the desired acceleration is the optimal solution.  Perhaps we need input from @lthall.  I'm pretty sure he wrote the comment, "// TODO: We need to also limit the _desired_accel" in AC_Loiter.cpp.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -640,13 +640,29 @@ void ModeGuided::set_desired_velocity_with_accel_and_fence_limits(const Vector3f
 #if AC_AVOID_ENABLED
     // limit the velocity to prevent fence violations
-    copter.avoid.adjust_velocity(pos_control->get_pos_xy_p().kP(), pos_control->get_max_accel_xy(), curr_vel_des, G_Dt);
+    Vector2f rel_save_location_cm;
```
"safe" instead of "save" I think

rmackay9 commented on libraries/AC_WPNav/AC_Loiter.cpp:
```diff
@@ -305,12 +305,22 @@ void AC_Loiter::calc_desired_velocity(float nav_dt)
 
     // Limit the velocity to prevent fence violations
     // TODO: We need to also limit the _desired_accel
+    Vector2f rel_save_location_cm;
```
"safe" instead of "save" I think

rmackay9 commented on libraries/AC_WPNav/AC_Loiter.cpp:
```diff
@@ -305,12 +305,22 @@ void AC_Loiter::calc_desired_velocity(float nav_dt)
 
     // Limit the velocity to prevent fence violations
     // TODO: We need to also limit the _desired_accel
```
We can remove this TODO comment now I think.

IamPete1 replied:
fixed and added a few more comments to loiter

IamPete1 commented:
Removed a left over variable, and done some more testing, https://youtu.be/O5BtzPxXdoQ Traversing works OK, get very slow sideways response, either because we're zeroing acceleration constantly or its still using the avoidance speed. Also a a demo of what happens if you see the floor, don't think this has actually made that any-worse tho.

IamPete1 commented:
Removed a left over variable, and done some more testing, https://youtu.be/O5BtzPxXdoQ Traversing works OK, get very slow sideways response, either because we're zeroing acceleration constantly or its still using the avoidance speed. Also a a demo of what happens if you see the floor, don't think this has actually made that any-worse tho.

rishabsingh3003 commented:
Some more testing. Posting the links here for @rmackay9 
1. Fast moving walls sandwiching the vehicle - https://youtu.be/eQOFsJYrZbk
2. Same test, slowed down walls - https://www.youtube.com/watch?v=Pypw73l4eX8&feature=youtu.be

IamPete1 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -847,27 +943,34 @@ float AC_Avoid::distance_to_lean_pct(float dist_m)
-// returns the maximum positive and negative roll and pitch percentages (in -1 ~ +1 range) based on the proximity sensor
-void AC_Avoid::get_proximity_roll_pitch_pct(float &roll_positive, float &roll_negative, float &pitch_positive, float &pitch_negative)
+// returns the roll and pitch angle limits as a percentage (in -1 ~ +1 range) based on the proximity sensor
+Vector2f AC_Avoid::get_proximity_roll_pitch_pct()
```
Yeah, should be a none functions change, I was just confused by it, hopefully this makes is simpler, the min max stuff pops up again ;)

rishabsingh3003 commented:
Okay so I have had some more time to look at this now. Overall its looking nice and I can't wait to test this on my vehicle. But there are a few more issues that I have discovered while reviewing and simulating. Here are my observations:
1. With Copter, I kept hitting Floating Point exception randomly and I knew something was wrong but never had GDB on when it happened.
2. Rover Simple Avoidance is partial broken post this PR,  to reproduce, do this:
- Make a fence, set up a mission, switch to mode Auto (before arming)
- Arm in Auto, you'll see a Floating point exception 
- maybe we need a autotest to catch this in future?

3. GDB traced it to the division on line 384: 
`const Vector2f stopping_point = position_xy + desired_vel_cms*(get_stopping_distance(kP, accel_cmss, desired_speed)/desired_speed);
`
Here desired speed might become zero. This was originally checked in Master (Line 314), but removed by mistake I think (not sure if you had some rationale there) 

4. It's kind of tough to follow which changes are functional improvements and which are refactoring the code, since both are mixed in the commits here (but that's not really your fault... Just a little tough to track relevant changes for me as a newbie). 

Let me know how else I can help. I'll test this on real vehicle as soon as these points are addressed.

IamPete1 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -79,37 +87,76 @@ AC_Avoid::AC_Avoid()
+        distance_max.x = MAX(distance_max.x,safe_location_rel_cm.x);
+        distance_max.y = MAX(distance_max.y,safe_location_rel_cm.y);
+        distance_min.x = MIN(distance_min.x,safe_location_rel_cm.x);
+        distance_min.y = MIN(distance_min.y,safe_location_rel_cm.y);
```
done

IamPete1 commented on libraries/AC_Avoidance/AC_Avoid.cpp:
```diff
@@ -847,27 +943,34 @@ float AC_Avoid::distance_to_lean_pct(float dist_m)
+    float roll_positive = 0;
+    float roll_negative = 0;
+    float pitch_positive = 0;
+    float pitch_negative = 0;
```
Changed to vector2f

IamPete1 commented:
@rishabsingh3003 I have added some is_positive checks to desired speed hopefully that will fix the FPE on rover and copter (I have not tested). The issue is that we can no longer just return if there is no desired speed, we may still need to move back. I have also split the refactor out into a different commit (I appreciate its a bit late now, sorry)

IamPete1 commented:
fixed another FPE, now passing rover test locally, I think I have them all now

rmackay9 commented on libraries/AC_WPNav/AC_Loiter.cpp:
```diff
@@ -304,13 +304,26 @@ void AC_Loiter::calc_desired_velocity(float nav_dt)
+        _pos_control.set_max_accel_xy(_accel_cmss * 0.5);
+
+        Vector3f curr_pos = _inav.get_position();
+        _pos_control.set_xy_target(curr_pos.x + rel_safe_location_cm.x + desired_vel.x * nav_dt, curr_pos.y + rel_safe_location_cm.y + desired_vel.y * nav_dt);
```
I don't think we should need to set the xy_target.. instead we should be able to calculate a velocity vector away from the obstacle.

IamPete1 replied:
I started off trying to set the velocity, I never got it working that well. The information you have from the proximity is a distance to move back, rather than writing another position to velocity controller its better just to use the existing one. However I admit that there is room for improvement here in the switching from velocity commands in loiter to position for avoiding.

rmackay9 replied:
ok, I can imagine that we need to get the position modified as well but quietly updating the position target inside this method called, "calc_desired_velocity()" doesn't seem good.

IamPete1 commented:
closed by https://github.com/ArduPilot/ardupilot/pull/14819
